python/Ganga/GPIDev/Base/Proxy.py

```diff
@@ -561,16 +570,25 @@ def GPIProxyObjectFactory(_obj):
         from Ganga.Core.exceptions import GangaException
         raise GangaException("%s is NOT a Proxyable object" % type(_obj))
 
-    if hasattr(_obj, proxyObject):
-        return getattr(_obj, proxyObject)
+    ## This is defined within Objects.py, we could probably store this elsehere
+    ## (We probably do) but as this is guaranteed to be accessible for GangaObjects I will use this
+    this_class = getattr(stripProxy(_obj).__class__, proxyClass)
 
-    if not hasattr(_obj.__class__, proxyClass):
-        from Ganga.Core.exceptions import GangaException
-        raise GangaException("%s is NOT a Proxyable object" % type(_obj))
+    ## Would LOVE to avoid having to perform all of these checks to get or wrap a class in a proxy here but
+    ## multiple different BAD states are currently passed to this function
+    ## In future I would like to see this fixed elsewhere and reduce the logic in this section of code
+    ## As of Ganga 6.1.15 this is needed to put objects into a semi-coherent state
 
-    proxy_class = getattr(_obj.__class__, proxyClass)
+    if isType(_obj, this_class):
+        if isProxy(_obj):
+            return _obj
+        elif not isProxy(_obj):
+            pass
+    else:
+        pass
 
-    return proxy_class(_existing=_obj)
+    proxy_class = this_class(_proxy_impl_obj_to_wrap=_obj)
+    return proxy_class
```

Comment:
I'm happy to revert 154 over the proposed code at 582, I wasn't sure what this was doing.

I'm still not keen on getting the proxyClass in this way from an object which in principle doesn't need to know about it so I feel my comments stand but I'm happy to get the proxyClass instance this way for now. (I'd rather construct it as required here if needed, or get from a list of known proxy-class types).

If the object is of GangaObject type (we know it is from 560) it currently always has the proxyClass so testing for it feels a bit excessive. I'm also of the opinion that this should be something that this function is concerned with so I'm not wanting to add a reliance on getting things from inside a GangaObject here.
I'd rather just get the proxyClass otherwise than worry about raising an error when it doesn't exist (yet?).

3 doesn't change it simply is a string change which I hold no strong ties to, as long as the string is unlikely to be used by a developer on construction of a raw proxied object from 'in front' of the proxy.